### PR TITLE
[BREAKING] Change _slot and _offset to use dot in inline assembly

### PR DESCRIPTION
**Pull request by chriseth**
Fixes https://github.com/ethereum/solidity/issues/9345

Todo:
 - [x] documentation

### CODE REVIEW

File: test/libsolidity/syntaxTests/inlineAssembly/invalid/variable_declaration_suffix_offset.sol [hunk 1]
```diff
@@ -1,15 +1,11 @@
 contract C {
     function f() public pure {
         assembly {
-            let x_offset := 1
-            let x_slot := 1
-            let _offset := 1
```
Comment (chriseth) [inline test/libsolidity/syntaxTests/inlineAssembly/invalid/variable_declaration_suffix_offset.sol:6]: Using a `.` here would not parse.

File: test/libsolidity/syntaxTests/inlineAssembly/shadowing/name_clash_in_import.sol [hunk 2]
```diff
@@ -15,4 +15,4 @@ contract B {
 }
 // ----
 // DeclarationError 3859: (b:105-106): This declaration shadows a declaration outside the inline assembly block.
-// DeclarationError 3859: (b:128-131): The prefix of this declaration conflicts with a declaration outside the inline assembly block.
```
Comment (chriseth) [inline test/libsolidity/syntaxTests/inlineAssembly/shadowing/name_clash_in_import.sol:19]: I am actually not sure what the purpose of this error was - maybe in preparation for path resolution in inline assembly.

**Comment by axic** [comment]
I think it would be cleaner in terms of change review (especially judging test cases) if this change would be separated and first reviewing/merging the "disallow dot in user identifiers" change.

**Comment by chriseth** [comment]
Split out the first part into https://github.com/ethereum/solidity/pull/9291 @axic

File: test/libsolidity/syntaxTests/inlineAssembly/invalid/variable_declaration_suffix_offset.sol [hunk 3]
```diff
@@ -1,15 +1,11 @@
...
+            let x.slot := 1
         }
     }
 }
 // ----
-// DeclarationError 9155: (79-87): In variable declarations _slot and _offset can not be used as a suffix.
-// DeclarationError 9155: (109-115): In variable declarations _slot and _offset can not be used as a suffix.
-// DeclarationError 9155: (137-144): In variable declarations _slot and _offset can not be used as a suffix.
-// DeclarationError 9155: (166-171): In variable declarations _slot and _offset can not be used as a suffix.
+// DeclarationError 3927: (79-87): User-defined identifiers in inline assembly cannot contain '.'.
```
Comment (axic) [inline test/libsolidity/syntaxTests/inlineAssembly/invalid/variable_declaration_suffix_offset.sol:10]: Hmm, wouldn't the better error here be saying `x` is not found?

File: test/libsolidity/syntaxTests/inlineAssembly/storage_reference_empty_offset.sol [hunk 4]
```diff
@@ -1,9 +1,9 @@
...
     function f() public pure {
         assembly {
-            let x := _offset
+            let x := .offset
         }
     }
 }
 // ----
-// DeclarationError 4794: (84-91): In variable names _slot and _offset can only be used as a suffix.
+// ParserError 1856: (84-85): Literal or identifier expected.
```
Comment (axic) [inline test/libsolidity/syntaxTests/inlineAssembly/storage_reference_empty_offset.sol:9]: Why does the parser allows identifiers starting with `.` here?

Comment (chriseth) [reply test/libsolidity/syntaxTests/inlineAssembly/storage_reference_empty_offset.sol:9]: It does not - it says "identifier expected"

**Comment by chriseth** [comment]
Ready for review.

File: test/libsolidity/syntaxTests/inlineAssembly/storage_reference_empty_offset.sol (hunk 4 above)
Comment (axic) [reply test/libsolidity/syntaxTests/inlineAssembly/storage_reference_empty_offset.sol:9]: Right, did not look at the left saying "ParserError".

File: test/libsolidity/syntaxTests/inlineAssembly/invalid/variable_declaration_suffix_offset.sol (hunk 3 above)
Comment (axic) [reply test/libsolidity/syntaxTests/inlineAssembly/invalid/variable_declaration_suffix_offset.sol:10]: ping @chriseth

Comment (chriseth) [reply test/libsolidity/syntaxTests/inlineAssembly/invalid/variable_declaration_suffix_offset.sol:10]: If this were `x.offset := 1`, then yes, but how would you declare a new yul variable overwriting an existing solidity variable?

File: libsolidity/analysis/ReferencesResolver.cpp [hunk 5]
```diff
@@ -195,9 +196,10 @@ void ReferencesResolver::operator()(yul::FunctionDefinition const& _function)
 
 void ReferencesResolver::operator()(yul::Identifier const& _identifier)
 {
-	bool isSlot = boost::algorithm::ends_with(_identifier.name.str(), "_slot");
-	bool isOffset = boost::algorithm::ends_with(_identifier.name.str(), "_offset");
+	bool isSlot = boost::algorithm::ends_with(_identifier.name.str(), ".slot");
+	bool isOffset = boost::algorithm::ends_with(_identifier.name.str(), ".offset");
```
Comment (ekpyron) [inline libsolidity/analysis/ReferencesResolver.cpp:200]: By the way: Maybe at some point we could replace this by actually resolving anything with dots as a path structure and looking among ``stackSlots()`` in the last part - then we'd automatically get this as well as consistent versions for memory and calldata arrays, etc.
But that's a separate discussion, no need to do it right here, right now.
Ah yes, you even have comments for that, nice :+1:!

File: docs/Solidity.g4 [hunk 6]
```diff
@@ -306,7 +306,7 @@ assemblyBlock
   : '{' assemblyItem* '}' ;
 
 assemblyExpression
-  : assemblyCall | assemblyLiteral ;
+  : assemblyCall | assemblyLiteral | assemblyIdentifier ;
```
Comment (ekpyron) [inline docs/Solidity.g4:309]: Not exactly sure how these work wrt ordering, but we should maybe take care that ``x`` in ``if x`` now actually falls under ``assemblyIdentifier``, not ``assemblyCall``... maybe we should *require* the parenthesis in the ``assembly call``?
Also ``return`` in ``assemblyCall`` is not an expression... well... I guess this just needs quite some work, maybe too much to fix it in here...

**Comment by hrkrshnn** [comment]
I think this needs to be mentioned in 070-breaking-changes.rst https://github.com/ethereum/solidity/blob/breaking/docs/070-breaking-changes.rst

**Review by ekpyron — COMMENTED** [review]
Maybe we could retain one test for the old ``_slot``, ``_offset``? It might be nice to have a suggesetion to change this, in case there's no such identifier in the assembly... but maybe that'd be more trouble than it's worth...

File: libsolidity/analysis/TypeChecker.cpp [hunk 7]
```diff
@@ -724,7 +724,7 @@ bool TypeChecker::visit(InlineAssembly const& _inlineAssembly)
 				}
 				else if (requiresStorage)
 				{
-					m_errorReporter.typeError(6617_error, _identifier.location, "The suffixes _offset and _slot can only be used on non-constant storage variables.");
+					m_errorReporter.typeError(6617_error, _identifier.location, "The suffixes .offset and .slot can only be used on non-constant storage variables.");
```
Comment (hrkrshnn) [inline libsolidity/analysis/TypeChecker.cpp:770]: Is it still considered as suffixes? Maybe something like "The members "offset" and "slot" are only available for non-constant state variables."

Comment (mijovic) [reply libsolidity/analysis/TypeChecker.cpp:770]: This is interesting point. Maybe we should change to members, as it sounds ore natural for access with  `.`

**Review by mijovic — COMMENTED** [review]
Looks good, only small open question for using member instead of suffix for `.slot` and `.offset`

**Comment by chriseth** [comment]
Not sure about the member - we are not doing proper path resolving yet - what do you think @ekpyron ?

Added a check for `_slot` and `_offset`.

**Comment by ekpyron** [comment]
In the long run we should make it proper path resolution and properly resolve ``stackItems()`` in them. But if we don't do that here, talking about "suffices" until we change that, seems fine to me.

Another thing, though: while we're at breaking things here, do we want to go with the ``stackItems()`` idea in general moving forward? Then maybe we should change this for memory references as well and not resolve the name itself, but only ``arr.mpos``. Then switching to proper path resolution and actually using ``stackItems()`` as well as adding this for calldata references and all that would become non-breaking, right?

**Comment by ekpyron** [comment]
*With* the ``stackItems()`` idea we'd also get external function types for free. But it'd also break accessing contract addresses, i.e. it would be ``C c; assembly { let x := c.address }`` instead of ``let x := c``. But all of that seems reasonable to me...

Do you think we'd have the time to do all this properly for 0.7 :-)? Or at least go as far as to make doing it non-breaking (i.e. change memory references and contract references here)?

**Comment by chriseth** [comment]
@ekpyron I actually think this would be more confusing that helpful, especially for types that only have one stack slot. Also it wolud expose the names of the slots while this was just an internal name for now. I think this is too much for 0.7.

**Comment by chriseth** [comment]
Rebased and fixed test.

**Review by ekpyron — APPROVED** [review]

**Pull request merged**